design/pv_backup_info.md

```diff
@@ -42,21 +42,23 @@ The 1 version of `VolumeInfo` definition is:
 type VolumeInfoV1 struct {
     PVCName        string    // The PVC's name. The format should be <namespace-name>/<PVC-name>
     PVName         string    // The PV name.
-    BackupMethod   string    // The way the volume data is backed up. The valid value includes `VeleroNativeSnapshot`, `PodVolumeBackup`, `CSISnapshot` and `Skipped`.
-    SnapshotDataMovement bool // Whether the volume's snapshot data is moved to specified storage.
+    BackupMethod   string    // The way the volume data is backed up. The valid value includes `VeleroNativeSnapshot`, `PodVolumeBackup` and `CSISnapshot`.
+    SnapshotDataMoved bool   // Whether the volume's snapshot data is moved to specified storage.
 
+    Skipped         boolean   // Whether the Volume is skipped in this backup.
     SkippedReason   string    // The reason for the volume is skipped in the backup.
```

Comment:
what about error scenarios?
where we tried to take snapshot through 1 of the routes but it failed 
Will we populate it in skipreason?
If not should we maybe have another field for errors?

Reply:
My understanding is that "skippedReason" will be populated with the error message.

Reply:
The error scenario could be recorded by a `State` field, but there is no such field for this structure., and it's on purpose.

The SkipReason is not used for recording the volume backing-up error. It's used for filtered-out volumes. 

The consideration is the VolumeInfo metadata is used to record the volumes' information in the Velero process, and the async operations' changes are not included. Most of the Volume data backing up processes are time-consuming.
The reason that the async operations changes are not included is that if they are included, the Velero server needs to update the new metadata file during the backup finalizing state. It will make the process more complicated, and it will be hard to meet the metadata's data immutability requirement: https://velero.io/docs/v1.12/backup-reference/#limitation.

Reply:
I think I completely understand that currently this is meant for making the restore flow deterministic in some way.

Decoupling from this design / approach from now, I think similar to how we have backup/restore results to show warnings/ errors. Having the same here would be helpful. This can basically be shown during velero backup describe etc.
It would be more of a PV level backup result struct **. I am fine with not tackling it right now.**

Also on immutability - I think the doc is the opposite, we are not guaranteeing immutability in any way. So updating this file during async should be fine.

now on the other hand can you explain certain behaviours - 
Example - during backup a PVC went to CSI datamovement based backup. VolumeInfo reflects this information. Now let's say during async phase the datamovement fails, the backup will be marked partially failed. 


However when we restore from this backup, current behavior that I recall is that Restore will be marked Completed since the backup didn't even have the failed snapshots. However, now that this struct is our source of truth, we'll try to restore such volumes and potentially go to PartiallyFailed.

Or is this behaviour disparity handled in some way already?


I think if not maintaining the error message, there is value in keeping success/ failure of each snapshot through a particular route.

Reply:
I agree that there was no guarantee of immutability from Velero, but that was broken after the async operations were introduced. The concern is that introducing more changes that need to modify the metadata after it's created, will be hard to achieve immutability someday.

For the case that CSI backing up completes, but data upload fails, I can not confirm, but Velero should check the DataUpload's information. During restore, Velero will create a ConfigMap according to the DataUpload, then a DataDownload will be created according to the ConfigMap. I didn't find any logic to check the DataUpload's result, so I suppose the restore will also end with partial failure.

The existing Volume backup information, including DataUpload and PVB, still works after the VolumeInfo structure is introduced. How about if there is not enough information in the VolumeInfo, fall back to the previous way?

Reply:
My worry overall is that we are making this struct to be in some sense the source of truth of how a PV is backed up. It is just missing the truth about if it was backed up successfully or not. 
Falling back to existing information kind of defeats the purpose of making it the source of truth.

Reply:
After a discussion in the Slack channel, we agreed to move on with PR, and we will continue with this topic in the future design of this area.